### ISSUE TITLE
Unified REST APIs for working with Bitcoin

### ISSUE THREAD

**Issue by nud3l** [bob-collective/bob#55]
**Is your feature request related to a problem? Please describe.**
Protocols built on top of Bitcoin have various APIs available, but they are early stage and often lack documentation. Moreover, they are not well integrated with existing REST APIs like electrs.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by nakul1010** [comment]
Jotting down Currently known APIs pros/cons
## Mempool.space

[Link](https://mempool.space/docs/api/rest)
**Pros**
- Supports `testnet`, `mainnet`, `signet`. 
- Rest Apis are well documented with examples.
- Electrum RPC interface for signet. 

**Cons**
- No mentions of `brc20`/`ordinals` getter functions.
- No wallet integration mostly all getter functions.
## Blockchain.com 

[Link](https://www.blockchain.com/explorer/api)

**Pros**
- Native blockchain info Wallet integration available. 
- Almost all the rest of Apis are supported by mempool available.  

**Cons**
- No testnet support

## BlockCypher

[Link ](https://www.blockcypher.com/apis.html)

**Pros**
- Almost all the rest of Apis are supported by `mempool` and are available
- Can send raw tx to testnet. BTC `testnet3` supported. 
- Can be used to inscribe data when sending raw tx. But no doc is available. 

**Cons**
- I didn't like the doc structure. 

## Unisat 

[Link](https://docs.unisat.io/dev/unisat-wallet-api)
It's a wallet Api. 
Pros
- Well-documented APIs for brc20 and other inscriptions.

Cons
- No testnet support yet for some apis. 
- I don't know how they will index testnet ordinals. Currently, the mainnet ordinals and testnet ordinals list are identical. 
- Mainnet all ordinals not indexed
- Not free? 
- need unisat wallet integration to see the inscription. 

## Ideal Api

Link: We need to design. :) 

The API should support the following important functionalities

- Should work for `mainet` , `testnet`, `local deployment`. 
- Should support most of the getter functions mentioned in` mempool.space`. 
- If we want to support Ordinals
	- Should support creating an inscription and allowing it to be signed through different wallets that supports taproot and sigwit. 
	- Should support indexer. For local deployment the dev can run his own indexer. Something like [here](https://github.com/Next-DAO/brc20_indexer)
- Should support creating a raw tx.

## References


- [ordinals explanation with example](https://tara-annison.medium.com/a-comprehensive-explanation-of-ordinals-nfts-on-bitcoin-67b11868e74f) 
- [BRC-20 transaction decoded](https://lumjc12.medium.com/guide-to-brc-20-trna-214175f02061#:~:text=Sending%20a%20BRC%2D20%20transaction%20involves%20a%20two%2Dphase%20process,between%20OP_FALSE%20and%20OP_ENDIF%20opcodes.)
- [brc20 gitbook](https://domo-2.gitbook.io/brc-20-experiment/) 
- [ordinals inscription](https://ordinals.com/inscriptions)
- [Bitcoin addresses](https://bitcoin.design/guide/glossary/address/)
- [Why Indexer is required?](https://medium.com/alexgobtc/brc20-and-moving-towards-decentralized-indexing-ce35c78410cb#:~:text=What%20is%20a%20BRC20%20index,balances%20is%20not%20self%2Devident.)

**Comment by nakul1010** [comment]
As discussed on Discord, the [following methods](https://github.com/interlay/electrs/blob/3325f1914b481a2f7b885d6365d2ad2bb902b280/src/rest.rs#L640) in electrs should be extended to handle

- `Method::GET`, `brc20_inscription` results will be given out in the form 
```json
{ 
  "p": "brc-20",
  "op": "mint",
  "tick": "ordi",
  "amt": "1"
}
```
- `Method::GET`, `pure_inscription`get pure inscription in text format.  
- `Method::GET`, `asset_tracking`get balance of asset on bob.  

cc: @nud3l and @ns212

**Comment by nud3l** [comment]
What about all the other APIs, e.g., the ones exposed by unisat to also inscribe BRC20 and ordinals plus the functions around BRC20 lists, holders, history, ...? https://docs.unisat.io/dev/open-api/brc20

I'd like to please have a more extensive proposal from you @nakul1010 @ns212 that covers:

1. The proposed new endpoints
2. The order in which you think those should be implemented

**Comment by nakul1010** [comment]
I was trying to add the APIs without changing much of the electrum structure and without adding an indexer. But if u want most of the functionality, then the following will be the flow. 

Changes in Electrum 
- Add `inscription_db: Option<Arc<RwLock<Inscriptions>>>` field
- For every block
   Get Block transactions -> Get witness data -> Parse inscriptions -> Store inscriptions inDB. 

For every GET API, lock the `<InscriptionDB>` field in read-only mode. The `<InscriptionDB>` field will act as an Indexer.

The following methods in the local environment can be accessible at endpoint `http://localhost:3003`. 

`Method::Get` `inscriptionUtxo/{address}`
- get the list of inscribed UTXO by address. 
- Output
```json
{

  "data": {
    "utxo": [
      {
        "address": "string",
        "codeType": 0,
        "height": 0,
        "idx": 0,
        "inscriptions": [
          {
            "inscriptionId": "string",
            "inscriptionNumber": 0,
            "isBRC20": true,
          }
        ],
        "satoshi": 10000,
        "scriptPk": "string",
        "scriptType": "string",
        "txid": "string",
      }
    ]
  }
}
```
----------

`Method::Get` `bestheight`
- Get the best block height of BRC20 data
- Output
```
{
  "data": {
    "height": 0,
    "total": 0
  }
}
```
----------

`Method::Get` `/inscriptions/<FROM>`
- Get the inscriptions in the given block to the latest block. 
- Output
```json
{

  "data": {
    "inscriptions": [
      {
		"inscriptionId": "string",
		"inscriptionNumber": 0,
		"isBRC20": true,
      },
      {
		"inscriptionId": "string",
		"inscriptionNumber": 1,
		"isBRC20": true,
      }
    ]
  }
}
```
----------

`Method::Get` `brc20/list`
- Output
```json
{
  "data": {
    "height": 813863,
    "total": 42905,
    "start": 0,
    "detail": [
      "ordi",
      "meme",
      "abcd",
    ]
  }
}
```
----------

`Method::Get` `brc20balance/<address>/<ticker>`
- Indexer require for this 
- Output
```json
{
  "data": {
    "ticker": "ordi",
    "overallBalance": "644000",
	"historyInscriptions": [
      {
        "data": null,
        "inscriptionNumber": 8196044,
        "inscriptionId": "decd853d761285b772b9dfcb4dc60d05df5cca07c46b5a6d6e9ef9a9f816e20di0",
        "confirmations": 0
      },
      {
        "data": null,
        "inscriptionNumber": 8122019,
        "inscriptionId": "5e79c838aabf8f13c80a504ad0166c6fdc9310568458f63fc0d7d59b32b5afc3i0",
        "confirmations": 0
      },
      {
        "data": null,
        "inscriptionNumber": 8122017,
        "inscriptionId": "a883d254d6c625413e33d68d2c5b73eb78c403d4c0521e7492931cafb5ef7c7fi0",
        "confirmations": 0
      },
      {
        "data": null,
        "inscriptionNumber": 8122018,
        "inscriptionId": "e1a2b505460926764e322f154a814d3624e7d6c33a6172fb94133611f77301a8i0",
        "confirmations": 0
      },
      {
        "data": null,
        "inscriptionNumber": 8196043,
        "inscriptionId": "8380dbaf93ba6b441d12633920d692864dddc8a574fa4818039ad20064ceb451i0",
        "confirmations": 0
      },
      {
        "data": null,
        "inscriptionNumber": 8076373,
        "inscriptionId": "bc00977dbd87fb22f54f5c1b4eba781991243ffefebfbe639db45983246dbfe5i0",
        "confirmations": 0
      },
      {
        "data": null,
        "inscriptionNumber": 8076372,
        "inscriptionId": "9115ce42d15d1ffb878a6af1077bde8a2da765f3e299c3fd915a5b353230a812i0",
        "confirmations": 0
      }
    ]
  }
}

```
----------

`Method::Get` `inscriptions/<address>/<ticker>`
- get inscriptions by account
- Output
```json
{
  "code": 0,
  "msg": "ok",
  "data": {
    "cursor": 0,
    "total": 29,
    "totalConfirmed": 29,
    "totalUnconfirmed": 0,
    "totalUnconfirmedSpend": 0,
    "utxo": [
      {
        "txid": "0328b6bb6d749cf2d2535eb5cef939a0f5db5bdf4540b80ff773c3a85888c2bb",
        "satoshi": 546,
        "scriptType": "0020",
        "inscriptions": [
          {
            "inscriptionNumber": 22818979,
            "inscriptionId": "a6cba6a3bd2cde4b7a600638838de5387aeb4f061968fb894f0cf14e0ee3b0edi0",
            "isBRC20": true
          }
        ]
      },
      {
        "txid": "c42584248fb93e0b8e51db8344d53cdaa3eb12e5e73e6ae72d855d23aa1a9b0b",
        "satoshi": 546,
        "inscriptions": [
          {
            "inscriptionNumber": 23338832,
            "inscriptionId": "9774f5827c242962d3f12f599221370f9e9d8b121605e6e29e6a32863717d96ei0",
            "isBRC20": true
          }
        ]
      },
    ]
  }
}
```

----------

- Need to research for the following Post methods in much more depth. But the input/outputs should not change. It will also depend upon user wallet it should support segwit and taproot addresses.  

`Method::Post` `create_inscriptions/<address>/<json_string>`

```json
{
  "code": 0,
  "msg": "ok",
  "data": {
	"Inscriptions": [
	   {
		   "inscriptionId": "9774f5827c242962d3f12f599221370f9e9d8b121605e6e29e6a32863717d96ei0",
		   "index": 0
	   },
	   {
		   "inscriptionId": "9774f5827c242962d3f12f599221370f9e9d8b121605e6e29e6a32863717d96ei0",
		   "index":1
	   },
	   {
		   "inscriptionId": "9774f5827c242962d3f12f599221370f9e9d8b121605e6e29e6a32863717d96ei0",
		   "index": 2
	   }
	]
  }
}
```

----------
`Method::Post` `create_batch_inscriptions/<address>/<json_string>/`

----------

`Method::Post` `brc20/deploy/<address>/<amount>`
- JSON input
```Json
{
  "receiveAddress": "",
  "feeRate": 1,
  "outputValue": 546,
  "devAddress": "",
  "devFee": 0,
  "brc20Ticker": "",
  "brc20Max": "",
  "brc20Limit": ""
}
```
- JSON Output
```json
{
  "code": 0,
  "msg": "string",
  "data": {
    "amount": 3000,
    "paidAmount": 0,
	"status": "pending",
    "feeRate": 0,
    "minerFee": 0,
    "files": [
      {
        "filename": "10000.sats",
        "inscriptionId": "",
        "status": "pending"
      }
    ]
  }
}
```
----------

`Method::Post` `brc20/transfer/<address>/<amount>`
- JSON Input
```json
{
  "receiveAddress": "",
  "feeRate": 1,
  "outputValue": 546,
  "devAddress": "",
  "devFee": 0,
  "brc20Ticker": "",
  "brc20Max": "",
  "brc20Limit": ""
}
```
- JSON Output
```json
{
  "code": 0,
  "msg": "string",
  "data": {
    "amount": 3000,
	"status": "pending",
    "paidAmount": 0,
    "feeRate": 0,
    "minerFee": 0,
    "files": [
      {
        "filename": "10000.sats",
        "inscriptionId": "",
        "status": "pending"
      }
    ]
  }
}
```
----------

`Method::Post` `brc20/mint/<address>/<amount>`

- JSON Input
```json
{
  "receiveAddress": "",
  "feeRate": 1,
  "outputValue": 546,
  "devAddress": "",
  "devFee": 0,
  "brc20Ticker": "",
  "brc20Max": "",
  "brc20Limit": ""
}
```
- JSON Output
```json
{
  "code": 0,
  "msg": "string",
  "data": {
    "amount": 3000,
    "paidAmount": 0,
	"status": "pending",
    "feeRate": 0,
    "minerFee": 0,
    "files": [
      {
        "filename": "10000.sats",
        "inscriptionId": "",
        "status": "pending"
      }
    ]
  }
}
```

----------
## Still Unknowns
- Should work for `regtest` as electrs will start from the genesis block. 
- For `testnet` and `mainnet`, if anyone tries to access the getter functions it will fail likely since inscriptions won't be found. 
- The unisat APIs are not open source so need to research POST methods. 

----------
## Proposal to implement

- Understanding [ord codebase](https://github.com/ordinals/ord/tree/master) completely. 
- Focusing on parsing each Block transaction, and creating `inscription_db` in electrs. 
- Implementing and testing all getter methods. 
- Implementing Post methods.

**Comment by nakul1010** [comment]
Closed in #73
